org.osgi.test.common/src/main/java/org/osgi/test/common/CloseableBundleContext.java

```diff
@@ -0,0 +1,239 @@
+package org.osgi.test.common;
+/*
+ * Copyright (c) OSGi Alliance (2019). All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import java.io.File;
+import java.io.InputStream;
+import java.util.Collection;
+import java.util.Dictionary;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.osgi.framework.Bundle;
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.BundleException;
+import org.osgi.framework.BundleListener;
+import org.osgi.framework.Filter;
+import org.osgi.framework.FrameworkListener;
+import org.osgi.framework.InvalidSyntaxException;
+import org.osgi.framework.ServiceFactory;
+import org.osgi.framework.ServiceListener;
+import org.osgi.framework.ServiceObjects;
+import org.osgi.framework.ServiceReference;
+import org.osgi.framework.ServiceRegistration;
+
+public class CloseableBundleContext implements AutoCloseable, BundleContext {
+
+	private final BundleContext					bundleContext;
+	private final Set<ServiceRegistration<?>>	regs		= ConcurrentHashMap.newKeySet();
+	private final Set<Object>					listeners	= ConcurrentHashMap.newKeySet();
```

Comment:
listeners must be identity based and you must use separate collections per listener type (framework, bundle, service) since the same object can implement all the listener types yet only be registered for one type.

Reply:
done!